File: src/engine/discovery.js
"use strict";

const fs = require("fs");
const path = require("path");
const yaml = require("js-yaml");
const schema = require("./schema");

const JOB_SUFFIX = ".job.yml";
const DEFAULTS_FILENAME = "_defaults.yml";
const ALLOWED_DEFAULT_KEYS = new Set([
  "timezone",
  "enabled",
  "retries",
  "retry_backoff_seconds",
  "timeout_seconds",
  "misfire_policy",
  "misfire_cap",
  "history_limit",
  "history_retention_days",
  "failure_policy",
  "notify",
]);

class DiscoveryError extends Error {}

function loadYaml(filePath) {
  const data = yaml.load(fs.readFileSync(filePath, "utf8")) || {};
  if (data === null || typeof data !== "object" || Array.isArray(data)) {
    throw new DiscoveryError(`${filePath}: expected a YAML mapping at the top level`);
  }
  return data;
}

function shallowMerge(base, override) {
  const result = { ...base };
  for (const [key, value] of Object.entries(override)) {
    if (
      value &&
      typeof value === "object" &&
      !Array.isArray(value) &&
      result[key] &&
      typeof result[key] === "object" &&
      !Array.isArray(result[key])
    ) {
      result[key] = { ...result[key], ...value };
    } else {
      result[key] = value;
    }
  }
  return result;
}

function walk(dir) {
  if (!fs.existsSync(dir)) return [];
  const out = [];
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const entryPath = path.join(dir, entry.name);
    if (entry.isDirectory()) out.push(...walk(entryPath));
    if (entry.isFile() && entry.name.endsWith(JOB_SUFFIX)) out.push(entryPath);
  }
  return out.sort();
}

function defaultsChainDirs(jobsRoot, jobDir) {
  const root = fs.realpathSync(jobsRoot);
  const dir = fs.realpathSync(jobDir);
  const rel = path.relative(root, dir);
  if (rel.startsWith("..") || path.isAbsolute(rel)) {
    throw new DiscoveryError(`${dir} is not under jobs root ${root}`);
  }
  const parts = rel ? rel.split(path.sep) : [];
  const dirs = [root];
  let cur = root;
  for (const part of parts) {
    cur = path.join(cur, part);
    dirs.push(cur);
  }
  return dirs;
}

function loadDefaultsChain(jobsRoot, jobDir) {
  let merged = {};
  for (const dir of defaultsChainDirs(jobsRoot, jobDir)) {
    const defaultsPath = path.join(dir, DEFAULTS_FILENAME);
    if (!fs.existsSync(defaultsPath)) continue;
    const raw = loadYaml(defaultsPath);
    const unknown = Object.keys(raw).filter((key) => !ALLOWED_DEFAULT_KEYS.has(key));
    if (unknown.length) {
      throw new DiscoveryError(`${defaultsPath}: unknown default key(s) ${JSON.stringify(unknown.sort())}`);
    }
    merged = shallowMerge(merged, raw);
  }
  return merged;
}

function discoverJobs(jobsRoot) {
  if (!fs.existsSync(jobsRoot)) return [];
  const jobs = [];
  const errors = [];
  for (const jobFile of walk(jobsRoot)) {
    try {
      const raw = loadYaml(jobFile);
      const defaults = loadDefaultsChain(jobsRoot, path.dirname(jobFile));
      const merged = shallowMerge(defaults, raw);
      schema.validateMerged(merged, jobFile);
      jobs.push(schema.buildJob(merged, jobFile));
    } catch (err) {
      if (err instanceof DiscoveryError || err instanceof schema.ValidationError) {
        errors.push(err.message);
      } else {
        errors.push(`${jobFile}: ${err.message}`);
      }
    }
  }
  if (errors.length) throw new DiscoveryError(errors.join("\n"));
  schema.checkDuplicateIds(jobs);
  return jobs;
}

module.exports = {
  JOB_SUFFIX,
  DEFAULTS_FILENAME,
  DiscoveryError,
  loadYaml,
  shallowMerge,
  discoverJobs,
};

File: engine/discovery.py
from __future__ import annotations

from pathlib import Path

import yaml

from . import schema
from .models import Job

JOB_SUFFIX = ".job.yml"
DEFAULTS_FILENAME = "_defaults.yml"

# Keys a _defaults.yml may set. Deliberately excludes id/schedule/type/name/http/script
# so identity and per-job specifics can never leak in through inheritance.
ALLOWED_DEFAULT_KEYS = {
    "timezone",
    "enabled",
    "retries",
    "retry_backoff_seconds",
    "timeout_seconds",
    "misfire_policy",
    "misfire_cap",
    "history_limit",
    "notify",
}


class DiscoveryError(Exception):
    """Raised for filesystem/YAML-level problems. May bundle multiple messages."""


def _load_yaml(path: Path) -> dict:
    with open(path) as f:
        data = yaml.safe_load(f) or {}
    if not isinstance(data, dict):
        raise DiscoveryError(f"{path}: expected a YAML mapping at the top level")
    return data


def _shallow_merge(base: dict, override: dict) -> dict:
    result = dict(base)
    for k, v in override.items():
        if isinstance(v, dict) and isinstance(result.get(k), dict):
            result[k] = {**result[k], **v}
        else:
            result[k] = v
    return result


def _defaults_chain_dirs(jobs_root: Path, job_dir: Path) -> list[Path]:
    """Directories from jobs_root down to job_dir, inclusive, shallow -> deep."""
    job_dir = job_dir.resolve()
    jobs_root = jobs_root.resolve()
    if jobs_root not in job_dir.parents and job_dir != jobs_root:
        raise DiscoveryError(f"{job_dir} is not under jobs root {jobs_root}")
    dirs = [job_dir]
    cur = job_dir
    while cur != jobs_root:
        cur = cur.parent
        dirs.append(cur)
    dirs.reverse()
    return dirs


def _load_defaults_chain(jobs_root: Path, job_dir: Path) -> dict:
    merged: dict = {}
    for d in _defaults_chain_dirs(jobs_root, job_dir):
        defaults_path = d / DEFAULTS_FILENAME
        if defaults_path.exists():
            raw = _load_yaml(defaults_path)
            unknown = set(raw.keys()) - ALLOWED_DEFAULT_KEYS
            if unknown:
                raise DiscoveryError(
                    f"{defaults_path}: unknown default key(s) {sorted(unknown)}"
                )
            merged = _shallow_merge(merged, raw)
    return merged


def discover_jobs(jobs_root: Path) -> list[Job]:
    """Walk jobs_root for *.job.yml files, merge _defaults.yml chains, validate,
    and return fully-built Job objects. Raises DiscoveryError/schema.ValidationError
    with every problem found bundled into one message, not just the first."""
    jobs_root = Path(jobs_root)
    if not jobs_root.exists():
        return []

    job_files = sorted(jobs_root.rglob(f"*{JOB_SUFFIX}"))
    jobs: list[Job] = []
    errors: list[str] = []

    for jf in job_files:
        try:
            raw = _load_yaml(jf)
            defaults = _load_defaults_chain(jobs_root, jf.parent)
            merged = _shallow_merge(defaults, raw)
            schema.validate_merged(merged, jf)
            jobs.append(schema.build_job(merged, jf))
        except (DiscoveryError, schema.ValidationError) as e:
            errors.append(str(e))

    if errors:
        raise DiscoveryError("\n".join(errors))

    schema.check_duplicate_ids(jobs)
    return jobs
